### PR TITLE
feat: Frames Client

### PR DESCRIPTION
**Pull request by alexrisch**
## Introduction 📟
Issue: https://github.com/xmtp/xmtp-ios/issues/298

## Purpose ℹ️ 
Allow for native clients to utilize logic for Frames Client

### CODE REVIEW

File: Sources/XMTPiOS/Frames/FramesClient.swift [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+class FramesClient {
+    var xmtpClient: Client
+    var proxy: OpenFramesProxy
+
+    init(xmtpClient: Client, proxy: OpenFramesProxy? = nil) {
+        self.xmtpClient = xmtpClient
+        self.proxy = proxy ?? OpenFramesProxy()
+    }
+
+    func signFrameAction(inputs: FrameActionInputs) async throws -> FramePostPayload {
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesClient.swift:23]: I think you need to add public on the front of this for it actually be accessible.
```suggestion
    public func signFrameAction(inputs: FrameActionInputs) async throws -> FramePostPayload {
```

File: Sources/XMTPiOS/Frames/FramesClient.swift [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import LibXMTP
+
+public typealias FrameActionBody = Xmtp_MessageContents_FrameActionBody
+public typealias FrameAction = Xmtp_MessageContents_FrameAction
+
+class FramesClient {
+    var xmtpClient: Client
+    var proxy: OpenFramesProxy
+
+    init(xmtpClient: Client, proxy: OpenFramesProxy? = nil) {
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesClient.swift:18]: Same for the init if you want people to be able to access it.
```suggestion
    public init(xmtpClient: Client, proxy: OpenFramesProxy? = nil) {
```

File: Sources/XMTPiOS/Frames/FramesClient.swift [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import Foundation
+import LibXMTP
+
+public typealias FrameActionBody = Xmtp_MessageContents_FrameActionBody
+public typealias FrameAction = Xmtp_MessageContents_FrameAction
+
+class FramesClient {
+    var xmtpClient: Client
+    var proxy: OpenFramesProxy
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesClient.swift:16]: If you want either of these reachable externally make sure to add public to them as well.

File: Sources/XMTPiOS/Frames/FramesClient.swift [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+//  Created by Alex Risch on 3/28/24.
+//
+
+import Foundation
+import LibXMTP
+
+public typealias FrameActionBody = Xmtp_MessageContents_FrameActionBody
+public typealias FrameAction = Xmtp_MessageContents_FrameAction
+
+class FramesClient {
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesClient.swift:14]: Do you plan on this being extended? If not probably safe to make the client final.

```suggestion
public final class FramesClient {
```

**Review by nplasterer — COMMENTED** [review]
I'm not totally sure how this is going to be used in a client. But from the test it looks like you're probably going to want to make a bunch of stuff public.

File: Sources/XMTPiOS/Frames/FramesClient.swift (hunk 4 above)
Comment (alexrisch) [reply Sources/XMTPiOS/Frames/FramesClient.swift]: I imagine someone _could_ want to extend it, not sure of the why though

**Review by neekolas — APPROVED** [review]
@alexrisch this looks fantastic! The fact that the E2E test works against a real frame gives me confidence this is doing all the signature stuff right.

File: Sources/XMTPiOS/Frames/FramesClient.swift [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        case .group(let groupInputs):
+            let combined = groupInputs.groupId + groupInputs.groupSecret
+            let digest = sha256(input: combined)
+            return digest.base64EncodedString()
+        case .dm(let dmInputs):
+            guard let conversationTopic = dmInputs.conversationTopic else {
+                throw InvalidArgumentsError()
+            }
+            guard let combined = (conversationTopic.lowercased() + dmInputs.participantAccountAddresses.map { $0.lowercased() }.sorted().joined()).data(using: .utf8) else {
+                throw InvalidArgumentsError()
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesClient.swift:94]:
```suggestion
                throw FramesClientError.missingConversationTopic
```

File: Sources/XMTPiOS/Frames/FramesErrors.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        switch self {
+        case .customError(_, let status):
+            return status
+        }
+    }
+}
+
+class InvalidArgumentsError: Error {
+    
+}
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesErrors.swift:30]: Do you need this?

File: Sources/XMTPiOS/Frames/FramesTypes.swift [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    enum CodingKeys: CodingKey {
+        case action, target, label
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let action = try container.decode(String.self, forKey: .action)
+        guard let target = try container.decodeIfPresent(String.self, forKey: .target) else {
+            throw InvalidArgumentsError()
```
Comment (nplasterer) [inline Sources/XMTPiOS/Frames/FramesTypes.swift:26]: Same here for better error messages.

**Review by nplasterer — APPROVED** [review]

**Pull request merged**